[Boost] Add deactivate hook to Cache Module

donnchawp opened the pull request:
This PR adds a deactivate hook and code to cleanup when Boost is deactivated.

It does 3 things:
* Removes the WP_CACHE from wp-config.php
* Deletes advanced-cache.php if it's the Boost one.
* Deletes the wp-content/boost-cache directory and sub directories to remove any cached files.

Something to consider - when we add a config file in wp-content/boost-cache/ it should be left there when the plugin is deactivated. If the plugin is uninstalled, then it should be deleted.

Fixes [#35204](https://github.com/Automattic/jetpack/issues/35204)

## Proposed changes:
* Adds new deactivate function to the Page_Cache class.
* Adds a Boost_Cache_Utils class that contains the delete_directories() function. This will be shared with the caching engine, eventually.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
pc9hqz-2rA-p2

## Does this pull request change what data or activity we track or use?
No

## Testing instructions:
* Patch plugin with this PR.
* Make sure that caching is activated and cache files are in wp-content/boost-cache/cache/
* Deactivate the plugin on the Plugins page.
* Check wp-config.php for WP_CACHE.
* Check if wp-content/advanced-cache.php and wp-content/boost-cache/ have been deleted.

donnchawp committed:
Added deactivate function, and Utils class

donnchawp committed:
changelog

dilirity commented on projects/plugins/boost/app/modules/cache/Boost_Cache_Utils.php:
```diff
@@ -0,0 +1,44 @@
+		}
+
+		// make sure that $dir is a directory inside WP_CONTENT . '/boost-cache/';
+		if ( self::is_boost_cache_directory( $dir ) === false ) {
+			return new \WP_Error( 'Invalid directory' );
+		}
```
Nice! When I saw this method, I immediately thought to myself "wouldn't it be nice if I were able to delete other stuff as well?" :D

dilirity commented on projects/plugins/boost/app/modules/cache/Boost_Cache_Utils.php:
```diff
@@ -0,0 +1,44 @@
+	 * @return bool
+	 */
+	public static function is_boost_cache_directory( $dir ) {
+		return strpos( $dir, WP_CONTENT_DIR . '/boost-cache' ) !== false;
```
Shenanigans, but `'/usr/local/htdocs/wp-content/boost-cache/../../test'` for `$dir` would be considered a valid boost directory 😅

donnchawp replied:
The realpath  call on line 12 should take care of that. It will convert that to /usr/local/htdocs/test/ - if that directory exists. Otherwise it returns false. I was considering reintroducing the sanitize function that uses deep_replace(). I may still do so I guess.

donnchawp replied:
I will sanitize it. It's not enough to rely on realpath() in delete_directory() because that returns false if the directory doesn't exist. It may happen in the future that we call is_boost_cache_directory() on a directory that we are yet to make.

dilirity commented on projects/plugins/boost/app/modules/cache/Boost_Cache_Utils.php:
```diff
@@ -0,0 +1,44 @@
+		}
+		rmdir( $dir ); // phpcs:ignore WordPress.WP.AlternativeFunctions.file_system_operations_rmdir
+
+		return true;
```
Maybe this should return the result of `rmdir`? We'd want to know if the directory was for some reason not deleted. Or not? I don't know what that would look like.

donnchawp replied:
I'll do that. I think I mentioned it elsewhere that this may return false if caching creates a file in the "cache" directory just before the rmdir() goes through.

dilirity commented on projects/plugins/boost/app/modules/cache/Page_Cache.php:
```diff
@@ -104,6 +114,55 @@ private function add_wp_cache_define() {
+		}
+
+		$content = file_get_contents( $advanced_cache_filename ); // phpcs:ignore WordPress.WP.AlternativeFunctions.file_get_contents_file_get_contents
+		if ( strpos( $content, 'Boost Cache Plugin 0.1' ) !== false ) {
```
`Boost Cache Plugin 0.1` I think it would be good if we move this to a variable.

donnchawp replied:
Good idea, done!

dilirity commented on projects/plugins/boost/app/modules/cache/Page_Cache.php:
```diff
@@ -5,7 +5,17 @@
 class Page_Cache implements Pluggable, Is_Always_On {
+
+	public function __construct() {
+		add_action( 'jetpack_boost_deactivate', array( $this, 'deactivate' ) );
```
Currently, Boost does cleanup on uninstall - https://github.com/Automattic/jetpack/blob/751ad2715f20a257992321924b8e9e1759946c21/projects/plugins/boost/jetpack-boost.php#L225-L232

I think it makes sense to cleanup the cache then as well. Deactivation might be accidental (or a failed plugin update), so if the cache is large, deleting it would hurt.

I'm wondering if it doesn't make more sense to move everything to a dedicated `uninstall.php` later. It's going to be much cleaner.

donnchawp replied:
The Minify module has it's own uninstall too that cleans up. Do you mean a single uninstall.php for all of Boost? That would be useful if it called the cleanup process in each module, but then you're just replicating what register_uninstall_hook() does.
I'll register an uninstall_hook function for the moment. Maybe we can look at it again after.

dilirity replied:
Yeah. I meant a separate file for all of Boost. Wasn't aware that Minify has it's own uninstall. I guess we should streamline everything. But that's for later.

dilirity replied:
To put a conclusion to the discussion - we'll probably revisit Boost's uninstall process in the future. The changes done in this PR are enough for now.

dilirity commented on projects/plugins/boost/app/modules/cache/Page_Cache.php:
```diff
@@ -104,6 +114,55 @@ private function add_wp_cache_define() {
+	 * Fired when the plugin is deactivated.
+	 */
+	public function deactivate() {
+		$this->delete_advanced_cache(); // how do we handle errors?
+		$this->delete_wp_cache_constant(); // how do we handle errors?
```
While reviewing I was thinking the exact same thing - "What do we do about errors during boost removal?".

I don't think we need to handle anything. Just hope we clean things up, and if something goes wrong, there's not much we can do.

For example, wp-rocket [doesn't do anything](https://github.com/wp-media/wp-rocket/blob/develop/inc/Engine/WPRocketUninstall.php) if there's a problem with the cleanup.

donnchawp replied:
Removal probably works fine for most sites. Occasionally, I'd see complaints about WP Super Cache leaving behind files but that might be because people deleted the plugin directory. Not much we can do to stop them doing that.

donnchawp replied:
I just spotted where [WPSC handles this](https://github.com/Automattic/jetpack/blob/trunk/projects/plugins/super-cache/wp-cache.php#L3733), and it sends an email to the site owner if files have been left behind, giving instructions on what to do.

donnchawp replied:
I'm going to half-implement this. If I send the actual email, we'll end up making a monstrosity of a thing with proper Jetpack formatting and all that..

dilirity replied:
Let's leave out emails for now. We'll figure this out later.

dilirity commented on projects/plugins/boost/app/modules/cache/Boost_Cache_Utils.php:
```diff
@@ -0,0 +1,44 @@
+				wp_delete_file( $file );
+			}
+		}
+		rmdir( $dir ); // phpcs:ignore WordPress.WP.AlternativeFunctions.file_system_operations_rmdir
```
`rmdir` might throw a warning if there are still files in there (though we hopefully removed them a before that).

donnchawp replied:
I almost guarantee it will return warnings at some stage. Not for uninstall/deactivation, but on a cache directory being cleaned up by garbage collection. One process will try to delete the directory and another will be creating files in the same directory. It's one of the reasons I have to hide the warnings from most filesystem calls in WPSC!

donnchawp replied:
Hiding the warnings now.

dilirity commented on projects/plugins/boost/app/modules/cache/Page_Cache.php:
```diff
@@ -104,6 +114,55 @@ private function add_wp_cache_define() {
+	 * Deletes the WP_CACHE define from wp-config.php
+	 * @return WP_Error if an error occurred.
+	 */
+	public function delete_wp_cache_constant() {
```
Just thinking about this, but - what do you think about `cleanup_wp_config` instead? Up to you ^^

donnchawp replied:
I was about to make that change, but then I saw the function to add it is called, `add_wp_cache_define()`, so it's probably better to leave it the way it is.

dilirity commented on projects/plugins/boost/app/modules/cache/Page_Cache.php:
```diff
@@ -104,6 +114,55 @@ private function add_wp_cache_define() {
+	public function delete_wp_cache_constant() {
+		$lines = file( ABSPATH . 'wp-config.php' );
+		foreach ( $lines as $key => $line ) {
+			if ( strpos( $line, 'WP_CACHE' ) !== false ) {
```
This would also remove

```php
define( 'WP_CACHE_KEY_SALT', '...' );
```

and any other lines that mention WP_CACHE in a sentence.

I think we should also avoid modifying the file if there's no constant defined.

Also - if the define is on the same line as the opening php tag, things go nuts :D

We need to make sure we're removing only what we need.

donnchawp replied:
I'll copy the preg_match from `add_wp_cache_define()`, but we should be reasonably safe from having it on the same line as <?php. WPSC removes the entire line and, thankfully, reported problems with uninstalling the plugin are rare.

dilirity reviewed (changes requested):
A few edge-cases I managed to find below.

donnchawp committed:
Sanitize the directory path, removing chars/strings like ".."

donnchawp committed:
Return value of rmdir() when deleting directories

donnchawp committed:
move advanced-cache signature to a variable.

donnchawp committed:
Add uninstall hook that deletes boost-cache directory

donnchawp committed:
Check more carefully for WP_CACHE before removal.

donnchawp committed:
Add removal instructions array, for future use.

donnchawp committed:
Merge branch 'trunk' into add/deactivate_boost_cache

donnchawp committed:
Cleanup after merge and fix uninstall

Making some functions static meant I had to remove logging of errors.

donnchawp committed:
Move version string to static variable

donnchawp committed:
Hide warnings from rmdir(). They'll happen.

There's no way to stop another process creating a file in a cache
directory, unless you put a big fat mutex flag on it that slows
everything down.

dilirity reviewed (approved):
Thanks for addressing everything!

:shipit:
